src/main/java/frc/robot/commands/LiftClimbAndPull.java

```diff
@@ -1,57 +1,79 @@
+import java.util.Arrays;
+import java.util.List;
 
 import com.chaos131.swerve.BaseSwerveDrive;
 
 import edu.wpi.first.math.geometry.Pose2d;
 import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Transform2d;
 import edu.wpi.first.math.geometry.Translation2d;
 import edu.wpi.first.wpilibj2.command.Command;
 import frc.robot.Constants;
+import frc.robot.Constants.LiftConstants;
+import frc.robot.subsystems.Launcher;
 import frc.robot.subsystems.Lift;
+import frc.robot.subsystems.Vision;
 import frc.robot.subsystems.swerve.SwerveDrive;
 import frc.robot.util.FieldPose2024;
 
 public class LiftClimbAndPull extends Command {
     private Lift m_lift;
     private liftState state = liftState.moveIntoPosition;
     private BaseSwerveDrive m_swerveDrive;
+    private Pose2d m_closestChain;
+    private Pose2d m_startClimbPose;
+    private Vision m_vision;
+    private Launcher m_launcher;
 
     private enum liftState{
-        moveIntoPosition, raiseLift, moveForwards, pullDown
+        moveIntoPosition, visionMove, moveForwards, pullDown
     }
 
-    public LiftClimbAndPull (Lift lift, BaseSwerveDrive swerveDrive){
+    public LiftClimbAndPull (Lift lift, BaseSwerveDrive swerveDrive, Vision vision, Launcher launcher){
         m_lift = lift;
         m_swerveDrive = swerveDrive;
-        addRequirements(lift, swerveDrive);
+        m_vision = vision;
+        m_launcher = launcher;
+        addRequirements(lift, swerveDrive, launcher);
     };
 
     @Override
     public void initialize() {
     //   m_lift.moveToHeight(Constants.LiftConstants.StartClimbHeight);
-        double x = Math.cos(FieldPose2024.StageSource.getCurrentAlliancePose().getRotation().getRadians());
-        double y = Math.sin(FieldPose2024.StageSource.getCurrentAlliancePose().getRotation().getRadians());
-        Translation2d pose = FieldPose2024.StageSource.getCurrentAlliancePose().getTranslation().plus(new Translation2d (x, y));
-        m_swerveDrive.setTarget(new Pose2d(pose, FieldPose2024.StageSource.getCurrentAlliancePose().getRotation()));
+        List<Pose2d> Stages = Arrays.asList(new Pose2d[]{
+            FieldPose2024.StageSource.getCurrentAlliancePose(),
+            FieldPose2024.StageAmp.getCurrentAlliancePose(),
+            FieldPose2024.StageFar.getCurrentAlliancePose()});
+        Pose2d closestChain = m_swerveDrive.getPose().nearest(Stages);
+        m_closestChain = closestChain;
+        double x = Math.cos(closestChain.getRotation().getRadians());
+        double y = Math.sin(closestChain.getRotation().getRadians());
+        Translation2d pose = closestChain.getTranslation().plus(new Translation2d (x, y));
+        m_startClimbPose = new Pose2d(pose, closestChain.getRotation());
+        m_swerveDrive.setTarget(new Pose2d(pose, closestChain.getRotation()));
     }
 
   @Override
   public void execute() {
     System.out.println(state);
     if (state == liftState.moveIntoPosition) {
```

Comment:
Same thing here as FireIntoAmp.java's concerns. We should start in liftState.visionMove.